kong/plugins/aws-lambda/v4.lua

```diff
@@ -0,0 +1,232 @@
+  end
+  table.sort(q)
+  return table.concat(q, "&")
+end
+
+local function derive_signing_key(kSecret, date, region, service)
+  local kDate = hmac("AWS4" .. kSecret, date)
+  local kRegion = hmac(kDate, region)
+  local kService = hmac(kRegion, service)
+  local kSigning = hmac(kService, "aws4_request")
+  return kSigning
+end
+
+local function prepare_awsv4_request(tbl)
+  local domain = tbl.domain or "amazonaws.com"
+  local region = tbl.region
+  local service = tbl.service
+  local request_method = tbl.method
+  local canonicalURI = tbl.canonicalURI
+  local path = tbl.path
+  if canonicalURI == nil and path ~= nil then
+    canonicalURI = canonicalise_path(path)
+  elseif canonicalURI == nil or canonicalURI == "" then
+    canonicalURI = "/"
+  end
+  local canonical_querystring = tbl.canonical_querystring
+  local query = tbl.query
+  if canonical_querystring == nil and query ~= nil then
+    canonical_querystring = canonicalise_query_string(query)
+  end
+  local req_headers = tbl.headers or {}
+  local req_payload = tbl.body
+  local access_key = tbl.access_key
+  local signing_key = tbl.signing_key
+  local secret_key
+  if signing_key == nil then
+    secret_key = tbl.secret_key
+    if secret_key == nil then
+      return nil, "either 'signing_key' or 'secret_key' must be provided"
+    end
+  end
+  local timestamp = tbl.timestamp or os.time()
+  local tls = tbl.tls
+  if tls == nil then tls = true end
+  local port = tbl.port or (tls and 443 or 80)
+  local req_date = os.date("!%Y%m%dT%H%M%SZ", timestamp)
+  local date = os.date("!%Y%m%d", timestamp)
+
+  local host = service .. "." .. region .. "." .. domain
+  local host_header do -- If the "standard" port is not in use, the port should be added to the Host header
+    local with_port
+    if tls then
+      with_port = port ~= 443
+    else
+      with_port = port ~= 80
+    end
+    if with_port then
+      host_header = string.format("%s:%d", host, port)
+    else
+      host_header = host
+    end
+  end
+
+  local headers = {
+    ["X-Amz-Date"] = req_date;
+    Host = host_header;
+  }
+  local add_auth_header = true
+  for k, v in pairs(req_headers) do
+    k = k:gsub("%f[^%z-]%w", string.upper) -- convert to standard header title case
+    if k == "Authorization" then
+      add_auth_header = false
+    elseif v == false then -- don't allow a default value for this header
+      v = nil
+    end
+    headers[k] = v
+  end
+
+  -- Task 1: Create a Canonical Request For Signature Version 4
+  -- http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+  local canonical_headers, signed_headers do
+    -- We structure this code in a way so that we only have to sort once.
+    canonical_headers, signed_headers = {}, {}
+    local i = 0
+    for name, value in pairs(headers) do
+      if value then -- ignore headers with 'false', they are used to override defaults
+        i = i + 1
+        local name_lower = name:lower()
```

Comment:
using `lower` here, while just above we're using `upper` on a first glance not efficient